### PR TITLE
Controlplane endpoint validation

### LINKED ISSUE #455: Setting ControlPlaneEndpoint results in invalid endpoint value
/kind bug

**What steps did you take and what happened:**
When a `ControlPlaneEndpoint` in the cluster provider spec's cluster config is specified, the kubeconfig has an invalid endpoint value. For example:

#### `cluster.yaml`

```yaml
apiVersion: "cluster.k8s.io/v1alpha1"
kind: Cluster
metadata:
  name: ha-example
spec:
  clusterNetwork:
    services:
      cidrBlocks: ["100.64.0.0/13"]
    pods:
      cidrBlocks: ["100.96.0.0/11"]
    serviceDomain: "cluster.local"
  providerSpec:
    value:
      apiVersion: vsphere.cluster.k8s.io/v1alpha1
      kind: VsphereClusterProviderSpec
      server: ""
      username: ""
      password: ""
      clusterConfiguration:
        controlPlaneEndpoint: ha-example.elb.amazonaws.com:443
```

#### `machines.yaml`

```yaml
apiVersion: cluster.k8s.io/v1alpha1
kind: MachineList
items:
- apiVersion: cluster.k8s.io/v1alpha1
  kind: Machine
  metadata:
    name: "ha-example-controlplane-1"
    labels:
      cluster.k8s.io/cluster-name: "ha-example"
  spec:
    providerSpec:
      value:
        apiVersion: vsphere.cluster.k8s.io/v1alpha1
        kind: VsphereMachineProviderSpec
        datacenter: "SDDC-Datacenter"
        datastore: "WorkloadDatastore"
        resourcePool: "clusterapi"
        folder: "clusterapi"
        network:
          devices:
          - networkName: "sddc-cgw-network-6"
            gateway4: 192.168.6.1
            ipAddrs:
            - 192.168.6.20/24
            nameservers:
            - 1.1.1.1
            - 1.0.0.1
        numCPUs: 2
        memoryMiB: 2048
        diskGiB: 50
        template: "ubuntu-1804-kube-v1.13.6"
    versions:
      kubelet: "1.13.6"
      controlPlane: "1.13.6"
```

In the vSphere manager logs the following errors occur:

```shell
E0724 17:45:14.575457       1 actuator.go:217] [cluster-actuator]/cluster.k8s.io/v1alpha1/default/ha-example "msg"="target cluster is not ready" "error"="unable to get core client for target cluster: host must be a URL or a host:port pair: \"https://\"" 
```

The error is originating from [this](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/127436f4c979bf81404ca5d100c148a83a5b0d2a/pkg/cloud/vsphere/services/kubeclient/kubeclient.go#L39) call.

However, this [code block](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/127436f4c979bf81404ca5d100c148a83a5b0d2a/pkg/cloud/vsphere/services/kubeconfig/kubeconfig.go#L46-L58) in `kubeconfig.go` should prevent the endpoint in the stored kubeconfig secret from ever being `https://`.

Debugging should include:

* Adding logging to the `kubeconfig.go` block to emit the control plane endpoint at various locations
* A test around `kubeconfig.New` to verify the produced kubeconfig file has a valid endpoint based on variations of the ingress value.

**What did you expect to happen:**
For the resulting endpoint value to be valid.

**Environment:**

- Cluster-api-provider-vsphere version: `v0.4.0-alpha.0`

### PR DESCRIPTION
**Pull request by pickledrick**
**What this PR does / why we need it**:

Allows for validation and correct generation of non-schemed server endpoints in kubeconfig.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #455 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Fix server endpoint kubeconfig generation when non-schemed URL passed into cluster spec.
```

### CODE REVIEW

File: pkg/cloud/vsphere/services/kubeconfig/kubeconfig_test.go [hunk 1]
```diff
@@ -18,43 +18,77 @@ package kubeconfig_test
...
 				return
 			}
-			t.Log(yaml)
+			k := Kubeconfig{}
+			err = yaml.Unmarshal([]byte(y), &k)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if k.Clusters[0].Cluster.Server != tc.expected {
```
Comment (pickledrick) [inline pkg/cloud/vsphere/services/kubeconfig/kubeconfig_test.go]: Not sure if I should check for `len(k.Clusters) > 0` here?

Comment (akutz) [reply pkg/cloud/vsphere/services/kubeconfig/kubeconfig_test.go]: Hi @pickledrick,

I appreciate the PR, but I must advise you that a lot of the work you're doing will be part of other CAPI components in v1a2. Would you like to schedule some time to chat to make sure you're not spending time on areas that will be deprecated soon?

Comment (akutz) [reply pkg/cloud/vsphere/services/kubeconfig/kubeconfig_test.go]: That said, this bug has been driving me nuts, and I have to say thank you from the bottom of my heart for addressing it!

File: pkg/cloud/vsphere/services/kubeconfig/kubeconfig.go [hunk 2]
```diff
@@ -43,18 +43,21 @@ func New(clusterName, controlPlaneEndpoint string, caKeyPair v1alpha1.KeyPair) (
 		return "", errors.New("key not found in status")
 	}
 
-	controlPlaneEndpointURL, err := url.Parse(controlPlaneEndpoint)
-	if err != nil {
-		if controlPlaneEndpointURL, err = url.Parse("https://" + controlPlaneEndpoint); err != nil {
+	var server string
+	// https://github.com/golang/go/blob/master/src/net/url/url.go#L469
+	if !strings.Contains(controlPlaneEndpoint, "://") {
+		server = strings.Split(controlPlaneEndpoint, "/")[0]
```
Comment (akutz) [inline pkg/cloud/vsphere/services/kubeconfig/kubeconfig.go]: Hi @pickledrick,

Should `SplitN` be used here?

Comment (pickledrick) [reply pkg/cloud/vsphere/services/kubeconfig/kubeconfig.go]: Happy to update, also added test cases for deeper paths.

**Comment by akutz** [comment]
/close

No longer relevant in v1a2

**Pull request closed without merging**